File: Tests/ProgressorCodecTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

/// Byte-level fixtures for the Tindeq wire protocol.
///
/// These are hand-computed, not round-tripped through our own encoder: a decoder
/// tested only against its own encoder agrees with itself about the wrong byte
/// order. Endianness bugs are the dangerous class here because they produce
/// plausible-looking wrong numbers (a 20 kg pull reading as 4.6e-41 or 1.4e14)
/// rather than an obvious crash.
final class ProgressorCodecTests: XCTestCase {

    private func littleEndianBytes(_ value: UInt32) -> [UInt8] {
        [UInt8(truncatingIfNeeded: value),
         UInt8(truncatingIfNeeded: value >> 8),
         UInt8(truncatingIfNeeded: value >> 16),
         UInt8(truncatingIfNeeded: value >> 24)]
    }

    private func sampleBytes(kg: Float, micros: UInt32) -> [UInt8] {
        littleEndianBytes(kg.bitPattern) + littleEndianBytes(micros)
    }

    private func weightPacket(_ samples: [(Float, UInt32)]) -> Data {
        let payload = samples.flatMap { sampleBytes(kg: $0.0, micros: $0.1) }
        return Data([0x01, UInt8(payload.count)] + payload)
    }

    // MARK: - Little-endian float fixtures

    /// 20.0f is IEEE-754 0x41A00000, so on the wire (LE) it is 00 00 A0 41.
    func testSingleWeightSampleDecodesLittleEndianFloat() {
        // tag 1 (weight), length 8, float32 20.0, uint32 12500 µs
        let packet = Data([0x01, 0x08,
                           0x00, 0x00, 0xA0, 0x41,
                           0xD4, 0x30, 0x00, 0x00])
        let events = ProgressorCodec.decode(packet)
        XCTAssertEqual(events, [.sample(ForceSample(kg: 20.0, deviceMicros: 12_500))])
    }

    /// The device batches ~8 samples per notification at 80 Hz — every one of them
    /// must come out, in order.
    func testBatchedWeightPacketDecodesEverySampleInOrder() {
        var payload = Data()
        payload.append(contentsOf: [0x00, 0x00, 0x80, 0x3F, 0x00, 0x00, 0x00, 0x00])  // 1.0 kg @ 0 µs
        payload.append(contentsOf: [0x00, 0x00, 0xD0, 0x40, 0xD4, 0x30, 0x00, 0x00])  // 6.5 kg @ 12500 µs
        payload.append(contentsOf: [0x00, 0x00, 0xA0, 0x41, 0xA8, 0x61, 0x00, 0x00])  // 20.0 kg @ 25000 µs
        let packet = Data([0x01, UInt8(payload.count)]) + payload

        let events = ProgressorCodec.decode(packet)
        XCTAssertEqual(events, [
            .sample(ForceSample(kg: 1.0, deviceMicros: 0)),
            .sample(ForceSample(kg: 6.5, deviceMicros: 12_500, isBatchStart: false)),
            .sample(ForceSample(kg: 20.0, deviceMicros: 25_000, isBatchStart: false)),
        ])
    }

    func testWeightWindowAcceptsItsExactBoundaries() {
        XCTAssertEqual(ProgressorCodec.decode(weightPacket([(-10.0, 0)])),
                       [.sample(ForceSample(kg: -10.0, deviceMicros: 0))])
        XCTAssertEqual(ProgressorCodec.decode(weightPacket([(165.0, 12_500)])),
                       [.sample(ForceSample(kg: 165.0, deviceMicros: 12_500))])
    }

    func testWeightWindowRejectsTheNextFloatBeyondEachBoundary() {
        XCTAssertEqual(ProgressorCodec.decode(weightPacket([(Float(-10).nextDown, 0)])), [])
        XCTAssertEqual(ProgressorCodec.decode(weightPacket([(Float(165).nextUp, 12_500)])), [])
    }

    func testNonFiniteAndWireImpossibleWeightsAreDroppedPairByPair() {
        let packet = weightPacket([
            (.nan, 0),
            (.infinity, 12_500),
            (-.infinity, 25_000),
            (Float(-10).nextDown, 37_500),
            (5_000, 50_000),
            (20, 62_500),
        ])

        XCTAssertEqual(ProgressorCodec.decode(packet),
                       [.sample(ForceSample(kg: 20, deviceMicros: 62_500))],
                       "the first emitted sample keeps the notification boundary")
    }

    // MARK: - Command responses (tag 0 needs the pending command to disambiguate)

    func testBatteryResponseDecodesAsMillivolts() {
        // 4200 mV = 0x1068 → LE 68 10 00 00
        let packet = Data([0x00, 0x04, 0x68, 0x10, 0x00, 0x00])
        XCTAssertEqual(ProgressorCodec.decode(packet, answering: .getBatteryVoltage),
                       [.battery(millivolts: 4200)])
    }

    func testAppVersionResponseDecodesAsTrimmedASCII() {
        let packet = Data([0x00, 0x06]) + Data("1.2.3".utf8) + Data([0x00])
        XCTAssertEqual(ProgressorCodec.decode(packet, answering: .getAppVersion),
                       [.appVersion("1.2.3")])
    }

    /// The device never echoes which command it is answering, so with nothing
    /// pending the only honest decode is the raw payload — NOT a guess that four
    /// bytes must mean battery.
    func testTagZeroWithoutPendingCommandStaysRaw() {
        let packet = Data([0x00, 0x04, 0x68, 0x10, 0x00, 0x00])
        XCTAssertEqual(ProgressorCodec.decode(packet),
                       [.commandResponse(Data([0x68, 0x10, 0x00, 0x00]))])
    }

    func testLowPowerWarningDecodes() {
        XCTAssertEqual(ProgressorCodec.decode(Data([0x04, 0x00])), [.lowPowerWarning])
    }

    func testUnknownTagIsSurfacedRatherThanDropped() {
        let packet = Data([0x63, 0x02, 0xAB, 0xCD])
        XCTAssertEqual(ProgressorCodec.decode(packet),
                       [.unknown(tag: 0x63, payload: Data([0xAB, 0xCD]))])
    }

    // MARK: - Robustness

    /// Real BLE delivers short reads. A declared length running past the buffer must
    /// stop the walk, never read past the end and never trap.
    func testTruncatedPacketStopsCleanly() {
        let packet = Data([0x01, 0x10, 0x00, 0x00, 0xA0, 0x41])   // claims 16 bytes, has 4
        XCTAssertEqual(ProgressorCodec.decode(packet), [])
    }

    func testNonMultipleOfEightWeightBlockIsRejectedWhole() {
        // Length 12 = one whole pair plus half of another.
        let packet = Data([0x01, 0x0C,
                           0x00, 0x00, 0xA0, 0x41, 0xD4, 0x30, 0x00, 0x00,
                           0x00, 0x00, 0x80, 0x3F])
        XCTAssertEqual(ProgressorCodec.decode(packet),
                       [.unknown(tag: 1, payload: Data(packet.dropFirst(2)))])
    }

    func testEmptyWeightBlockIsRejected() {
        XCTAssertEqual(ProgressorCodec.decode(Data([0x01, 0x00])),
                       [.unknown(tag: 1, payload: Data())])
    }

    func testMalformedRFDPeakNeverMasqueradesAsACommandResponse() {
        for payload in [Data(), Data(repeating: 0xAA, count: 7), Data(repeating: 0xAA, count: 9)] {
            let packet = Data([0x02, UInt8(payload.count)]) + payload
            XCTAssertEqual(ProgressorCodec.decode(packet, answering: .getBatteryVoltage),
                           [.unknown(tag: 2, payload: payload)])
        }
    }

    func testRFDSeriesRequiresANonemptyWholeNumberOfPairs() {
        for payload in [Data(), Data(repeating: 0xAA, count: 7), Data(repeating: 0xAA, count: 9)] {
            let packet = Data([0x03, UInt8(payload.count)]) + payload
            XCTAssertEqual(ProgressorCodec.decode(packet),
                           [.unknown(tag: 3, payload: payload)])
        }

        let invalidForce = Data(sampleBytes(kg: .nan, micros: 12_500))
        XCTAssertEqual(ProgressorCodec.decode(Data([0x03, 0x08]) + invalidForce),
                       [.unknown(tag: 3, payload: invalidForce)])
    }

    func testFiveOneSecondTimestampDeltasSuppressTheWholeNotification() {
        let packet = weightPacket((0..<5).map { (20, UInt32($0) * 1_000_000) })
        XCTAssertEqual(ProgressorCodec.decode(packet), [])
    }

    func testNotificationSpanBudgetCoversPackedWeightBlocks() {
        let first = weightPacket([(20, 0)])
        let second = weightPacket([(20, 200_000)])
        XCTAssertEqual(ProgressorCodec.decode(first + second), [],
                       "each block is valid alone, but the notification span is not")
    }

    func testPackedMultiTLVPacketKeepsCompleteBlocksBeforeATruncatedTail() {
        let first = weightPacket([(20, 1_000)])
        let warning = Data([0x04, 0x00])
        let second = weightPacket([(21, 13_500)])
        let truncatedTail = Data([0x01, 0x08, 0x00])

        XCTAssertEqual(ProgressorCodec.decode(first + warning + second + truncatedTail), [
            .sample(ForceSample(kg: 20, deviceMicros: 1_000)),
            .lowPowerWarning,
            .sample(ForceSample(kg: 21, deviceMicros: 13_500, isBatchStart: false)),
        ])
    }

    func testEmptyAndHeaderOnlyPacketsDecodeToNothing() {
        XCTAssertEqual(ProgressorCodec.decode(Data()), [])
        XCTAssertEqual(ProgressorCodec.decode(Data([0x01])), [])
    }

    // MARK: - Timestamp wrap

    /// The device's µs clock is a UInt32, so it wraps every ~71.6 minutes. Naive
    /// subtraction would yield a huge negative interval and blow up a rep's timing;
    /// wrapping subtraction gives the true small delta.
    func testMicrosecondDeltaSurvivesUInt32Wrap() {
        let before = ForceSample(kg: 20, deviceMicros: 4_294_960_000)
        let after = ForceSample(kg: 20, deviceMicros: 50_000)
        XCTAssertEqual(after.microsSince(before), 57_296)
    }

    func testMicrosecondDeltaInTheNormalCase() {
        let a = ForceSample(kg: 20, deviceMicros: 1_000_000)
        let b = ForceSample(kg: 20, deviceMicros: 1_012_500)
        XCTAssertEqual(b.microsSince(a), 12_500)
    }

    // MARK: - Command encoding

    func testZeroPayloadCommandsEncodeAsABareOpcode() {
        XCTAssertEqual(ProgressorCommand.tare.encoded, Data([100]))
        XCTAssertEqual(ProgressorCommand.startWeightMeasurement.encoded, Data([101]))
        XCTAssertEqual(ProgressorCommand.stopWeightMeasurement.encoded, Data([102]))
        XCTAssertEqual(ProgressorCommand.enterSleep.encoded, Data([110]))
        XCTAssertEqual(ProgressorCommand.getBatteryVoltage.encoded, Data([111]))
    }

    func testOnlyQueryCommandsExpectAResponse() {
        XCTAssertTrue(ProgressorCommand.getBatteryVoltage.expectsResponse)
        XCTAssertTrue(ProgressorCommand.getAppVersion.expectsResponse)
        XCTAssertFalse(ProgressorCommand.tare.expectsResponse)
        XCTAssertFalse(ProgressorCommand.startWeightMeasurement.expectsResponse)
    }

    func testBatteryFractionClampsToTheCellRange() {
        XCTAssertEqual(ProgressorCodec.batteryFraction(millivolts: 4200), 1.0, accuracy: 0.001)
        XCTAssertEqual(ProgressorCodec.batteryFraction(millivolts: 3300), 0.0, accuracy: 0.001)
        XCTAssertEqual(ProgressorCodec.batteryFraction(millivolts: 3750), 0.5, accuracy: 0.001)
        XCTAssertEqual(ProgressorCodec.batteryFraction(millivolts: 5000), 1.0, accuracy: 0.001)
        XCTAssertEqual(ProgressorCodec.batteryFraction(millivolts: 1000), 0.0, accuracy: 0.001)
    }
}

File: android/engine/src/test/kotlin/run/nuri/getagrip/engine/ProgressorCodecTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import kotlin.math.nextDown
import kotlin.math.nextUp
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/// Byte-level fixtures for the Tindeq wire protocol.
///
/// These are hand-computed, not round-tripped through our own encoder: a decoder
/// tested only against its own encoder agrees with itself about the wrong byte
/// order. Endianness bugs are the dangerous class here because they produce
/// plausible-looking wrong numbers (a 20 kg pull reading as 4.6e-41 or 1.4e14)
/// rather than an obvious crash.
///
/// Translated from Tests/ProgressorCodecTests.swift.
class ProgressorCodecTests {

    private fun littleEndianBytes(value: UInt): ByteArray = bytes(
        (value and 0xFFu).toInt(),
        ((value shr 8) and 0xFFu).toInt(),
        ((value shr 16) and 0xFFu).toInt(),
        ((value shr 24) and 0xFFu).toInt(),
    )

    private fun sampleBytes(kg: Float, micros: UInt): ByteArray =
        littleEndianBytes(kg.toRawBits().toUInt()) + littleEndianBytes(micros)

    private fun weightPacket(samples: List<Pair<Float, UInt>>): ByteArray {
        var payload = ByteArray(0)
        for (sample in samples) payload += sampleBytes(kg = sample.first, micros = sample.second)
        return bytes(0x01, payload.size) + payload
    }

    private fun sample(kg: Double, micros: UInt, batchStart: Boolean = true) =
        ProgressorEvent.Sample(ForceSample(kg = kg, deviceMicros = micros, isBatchStart = batchStart))

    // MARK: - Little-endian float fixtures

    /// 20.0f is IEEE-754 0x41A00000, so on the wire (LE) it is 00 00 A0 41.
    @Test
    fun singleWeightSampleDecodesLittleEndianFloat() {
        // tag 1 (weight), length 8, float32 20.0, uint32 12500 µs
        val packet = bytes(
            0x01, 0x08,
            0x00, 0x00, 0xA0, 0x41,
            0xD4, 0x30, 0x00, 0x00,
        )
        val events = ProgressorCodec.decode(packet)
        assertEquals(listOf(sample(20.0, 12_500u)), events)
    }

    /// The device batches ~8 samples per notification at 80 Hz — every one of them
    /// must come out, in order.
    @Test
    fun batchedWeightPacketDecodesEverySampleInOrder() {
        val payload =
            bytes(0x00, 0x00, 0x80, 0x3F, 0x00, 0x00, 0x00, 0x00) +   // 1.0 kg @ 0 µs
                bytes(0x00, 0x00, 0xD0, 0x40, 0xD4, 0x30, 0x00, 0x00) +   // 6.5 kg @ 12500 µs
                bytes(0x00, 0x00, 0xA0, 0x41, 0xA8, 0x61, 0x00, 0x00)     // 20.0 kg @ 25000 µs
        val packet = bytes(0x01, payload.size) + payload

        val events = ProgressorCodec.decode(packet)
        assertEquals(
            listOf(
                sample(1.0, 0u),
                sample(6.5, 12_500u, batchStart = false),
                sample(20.0, 25_000u, batchStart = false),
            ),
            events,
        )
    }

    @Test
    fun weightWindowAcceptsItsExactBoundaries() {
        assertEquals(
            listOf(sample(-10.0, 0u)),
            ProgressorCodec.decode(weightPacket(listOf(-10.0f to 0u))),
        )
        assertEquals(
            listOf(sample(165.0, 12_500u)),
            ProgressorCodec.decode(weightPacket(listOf(165.0f to 12_500u))),
        )
    }

    @Test
    fun weightWindowRejectsTheNextFloatBeyondEachBoundary() {
        assertEquals(
            emptyList(),
            ProgressorCodec.decode(weightPacket(listOf((-10.0f).nextDown() to 0u))),
        )
        assertEquals(
            emptyList(),
            ProgressorCodec.decode(weightPacket(listOf(165.0f.nextUp() to 12_500u))),
        )
    }

    @Test
    fun nonFiniteAndWireImpossibleWeightsAreDroppedPairByPair() {
        val packet = weightPacket(
            listOf(
                Float.NaN to 0u,
                Float.POSITIVE_INFINITY to 12_500u,
                Float.NEGATIVE_INFINITY to 25_000u,
                (-10.0f).nextDown() to 37_500u,
                5_000.0f to 50_000u,
                20.0f to 62_500u,
            ),
        )

        assertEquals(
            listOf(sample(20.0, 62_500u)),
            ProgressorCodec.decode(packet),
            "the first emitted sample keeps the notification boundary",
        )
    }

    // MARK: - Command responses (tag 0 needs the pending command to disambiguate)

    @Test
    fun batteryResponseDecodesAsMillivolts() {
        // 4200 mV = 0x1068 → LE 68 10 00 00
        val packet = bytes(0x00, 0x04, 0x68, 0x10, 0x00, 0x00)
        assertEquals(
            listOf(ProgressorEvent.Battery(millivolts = 4200u)),
            ProgressorCodec.decode(packet, answering = ProgressorCommand.getBatteryVoltage),
        )
    }

    @Test
    fun appVersionResponseDecodesAsTrimmedASCII() {
        val packet = bytes(0x00, 0x06) + "1.2.3".toByteArray(Charsets.UTF_8) + bytes(0x00)
        assertEquals(
            listOf(ProgressorEvent.AppVersion("1.2.3")),
            ProgressorCodec.decode(packet, answering = ProgressorCommand.getAppVersion),
        )
    }

    /// The device never echoes which command it is answering, so with nothing
    /// pending the only honest decode is the raw payload — NOT a guess that four
    /// bytes must mean battery.
    @Test
    fun tagZeroWithoutPendingCommandStaysRaw() {
        val packet = bytes(0x00, 0x04, 0x68, 0x10, 0x00, 0x00)
        assertEquals(
            listOf(ProgressorEvent.CommandResponse(bytes(0x68, 0x10, 0x00, 0x00))),
            ProgressorCodec.decode(packet),
        )
    }

    @Test
    fun lowPowerWarningDecodes() {
        assertEquals(
            listOf(ProgressorEvent.LowPowerWarning),
            ProgressorCodec.decode(bytes(0x04, 0x00)),
        )
    }

    @Test
    fun unknownTagIsSurfacedRatherThanDropped() {
        val packet = bytes(0x63, 0x02, 0xAB, 0xCD)
        assertEquals(
            listOf(ProgressorEvent.Unknown(tag = 0x63, payload = bytes(0xAB, 0xCD))),
            ProgressorCodec.decode(packet),
        )
    }

    // MARK: - Robustness

    /// Real BLE delivers short reads. A declared length running past the buffer must
    /// stop the walk, never read past the end and never trap.
    @Test
    fun truncatedPacketStopsCleanly() {
        val packet = bytes(0x01, 0x10, 0x00, 0x00, 0xA0, 0x41)   // claims 16 bytes, has 4
        assertEquals(emptyList(), ProgressorCodec.decode(packet))
    }

    @Test
    fun nonMultipleOfEightWeightBlockIsRejectedWhole() {
        // Length 12 = one whole pair plus half of another.
        val packet = bytes(
            0x01, 0x0C,
            0x00, 0x00, 0xA0, 0x41, 0xD4, 0x30, 0x00, 0x00,
            0x00, 0x00, 0x80, 0x3F,
        )
        assertEquals(
            listOf(ProgressorEvent.Unknown(tag = 1, payload = packet.copyOfRange(2, packet.size))),
            ProgressorCodec.decode(packet),
        )
    }

    @Test
    fun emptyWeightBlockIsRejected() {
        assertEquals(
            listOf(ProgressorEvent.Unknown(tag = 1, payload = ByteArray(0))),
            ProgressorCodec.decode(bytes(0x01, 0x00)),
        )
    }

    @Test
    fun malformedRFDPeakNeverMasqueradesAsACommandResponse() {
        for (payload in listOf(ByteArray(0), ByteArray(7) { 0xAA.toByte() }, ByteArray(9) { 0xAA.toByte() })) {
            val packet = bytes(0x02, payload.size) + payload
            assertEquals(
                listOf(ProgressorEvent.Unknown(tag = 2, payload = payload)),
                ProgressorCodec.decode(packet, answering = ProgressorCommand.getBatteryVoltage),
            )
        }
    }

    @Test
    fun rfdSeriesRequiresANonemptyWholeNumberOfPairs() {
        for (payload in listOf(ByteArray(0), ByteArray(7) { 0xAA.toByte() }, ByteArray(9) { 0xAA.toByte() })) {
            val packet = bytes(0x03, payload.size) + payload
            assertEquals(
                listOf(ProgressorEvent.Unknown(tag = 3, payload = payload)),
                ProgressorCodec.decode(packet),
            )
        }

        val invalidForce = sampleBytes(kg = Float.NaN, micros = 12_500u)
        assertEquals(
            listOf(ProgressorEvent.Unknown(tag = 3, payload = invalidForce)),
            ProgressorCodec.decode(bytes(0x03, 0x08) + invalidForce),
        )
    }

    @Test
    fun fiveOneSecondTimestampDeltasSuppressTheWholeNotification() {
        val packet = weightPacket((0 until 5).map { 20.0f to it.toUInt() * 1_000_000u })
        assertEquals(emptyList(), ProgressorCodec.decode(packet))
    }

    @Test
    fun notificationSpanBudgetCoversPackedWeightBlocks() {
        val first = weightPacket(listOf(20.0f to 0u))
        val second = weightPacket(listOf(20.0f to 200_000u))
        assertEquals(
            emptyList(),
            ProgressorCodec.decode(first + second),
            "each block is valid alone, but the notification span is not",
        )
    }

    @Test
    fun packedMultiTLVPacketKeepsCompleteBlocksBeforeATruncatedTail() {
        val first = weightPacket(listOf(20.0f to 1_000u))
        val warning = bytes(0x04, 0x00)
        val second = weightPacket(listOf(21.0f to 13_500u))
        val truncatedTail = bytes(0x01, 0x08, 0x00)

        assertEquals(
            listOf(
                sample(20.0, 1_000u),
                ProgressorEvent.LowPowerWarning,
                sample(21.0, 13_500u, batchStart = false),
            ),
            ProgressorCodec.decode(first + warning + second + truncatedTail),
        )
    }

    @Test
    fun emptyAndHeaderOnlyPacketsDecodeToNothing() {
        assertEquals(emptyList(), ProgressorCodec.decode(ByteArray(0)))
        assertEquals(emptyList(), ProgressorCodec.decode(bytes(0x01)))
    }

    // MARK: - Timestamp wrap

    /// The device's µs clock is a UInt32, so it wraps every ~71.6 minutes. Naive
    /// subtraction would yield a huge negative interval and blow up a rep's timing;
    /// wrapping subtraction gives the true small delta.
    @Test
    fun microsecondDeltaSurvivesUInt32Wrap() {
        val before = ForceSample(kg = 20.0, deviceMicros = 4_294_960_000u)
        val after = ForceSample(kg = 20.0, deviceMicros = 50_000u)
        assertEquals(57_296u, after.microsSince(before))
    }

    @Test
    fun microsecondDeltaInTheNormalCase() {
        val a = ForceSample(kg = 20.0, deviceMicros = 1_000_000u)
        val b = ForceSample(kg = 20.0, deviceMicros = 1_012_500u)
        assertEquals(12_500u, b.microsSince(a))
    }

    // MARK: - Command encoding

    @Test
    fun zeroPayloadCommandsEncodeAsABareOpcode() {
        assertBytesEqual(bytes(100), ProgressorCommand.tare.encoded)
        assertBytesEqual(bytes(101), ProgressorCommand.startWeightMeasurement.encoded)
        assertBytesEqual(bytes(102), ProgressorCommand.stopWeightMeasurement.encoded)
        assertBytesEqual(bytes(110), ProgressorCommand.enterSleep.encoded)
        assertBytesEqual(bytes(111), ProgressorCommand.getBatteryVoltage.encoded)
    }

    @Test
    fun onlyQueryCommandsExpectAResponse() {
        assertTrue(ProgressorCommand.getBatteryVoltage.expectsResponse)
        assertTrue(ProgressorCommand.getAppVersion.expectsResponse)
        assertFalse(ProgressorCommand.tare.expectsResponse)
        assertFalse(ProgressorCommand.startWeightMeasurement.expectsResponse)
    }

    @Test
    fun batteryFractionClampsToTheCellRange() {
        assertEquals(1.0, ProgressorCodec.batteryFraction(4200u), 0.001)
        assertEquals(0.0, ProgressorCodec.batteryFraction(3300u), 0.001)
        assertEquals(0.5, ProgressorCodec.batteryFraction(3750u), 0.001)
        assertEquals(1.0, ProgressorCodec.batteryFraction(5000u), 0.001)
        assertEquals(0.0, ProgressorCodec.batteryFraction(1000u), 0.001)
    }
}
